yarn-project/circuits.js/src/utils/call_wasm.ts

```diff
@@ -11,19 +12,63 @@ export async function callAsyncWasm<T>(
 
   // Allocate memory for the input buffer and the pointer to the pointer to the output buffer
   const inputBufPtr = wasm.call('bbmalloc', inputBuf.length);
-  wasm.writeMemory(inputBufPtr, inputBuf);
-  const outputBufPtrPtr = wasm.call('bbmalloc', 4);
-
-  // Run and read outputs
-  const outputBufSize = await wasm.asyncCall(method, inputBufPtr, outputBufPtrPtr);
-  const outputBufPtr = uint8ArrayToNum(wasm.getMemorySlice(outputBufPtrPtr, outputBufPtrPtr + 4));
-  const outputBuf = Buffer.from(wasm.getMemorySlice(outputBufPtr, outputBufPtr + outputBufSize));
-  const output = outputType.fromBuffer(outputBuf);
-
-  // Free memory
-  wasm.call('bbfree', outputBufPtr);
-  wasm.call('bbfree', outputBufPtrPtr);
-  wasm.call('bbfree', inputBufPtr);
-
-  return output;
-}
+  wasm.writeMemory(inputBufPtr, inputBuf);  
+  if (method === 'merge_rollup__sim') {
+    // TODO: Remove this once base/root also return circuit failures
+    const outputBufSizePtr = wasm.call('bbmalloc', 4);
+    const outputBufPtrPtr = wasm.call('bbmalloc', 4);
+    // Run and read outputs
+    const circuitFailureBufPtr = await wasm.asyncCall(method, inputBufPtr, outputBufSizePtr, outputBufPtrPtr);
+    if (circuitFailureBufPtr == 0) {
+      // C++ returned a null pointer i.e. circuit didn't have an error
+      const outputBufSize = uint8ArrayToNum(wasm.getMemorySlice(outputBufSizePtr, outputBufSizePtr + 4));
+      const outputBufPtr = uint8ArrayToNum(wasm.getMemorySlice(outputBufPtrPtr, outputBufPtrPtr + 4));
+      const outputBuf = Buffer.from(wasm.getMemorySlice(outputBufPtr, outputBufPtr + outputBufSize));
+      const output = outputType.fromBuffer(outputBuf);
+
+      // Free memory
+      wasm.call('bbfree', circuitFailureBufPtr);
+      wasm.call('bbfree', outputBufPtr);
+      wasm.call('bbfree', outputBufPtrPtr);
+      wasm.call('bbfree', inputBufPtr);
+
+      return output;
+    } else {
+      // CircuitError struct is structured as:
+      // 1st 16 bits after the `circuitFailureBufPtr` - error code (enum uint16)
+      // Next 32 bits - error message size
+      // Next `error message size` bytes - error message.
+      // So need to first extract the error message size so we know how much memory to read for the entire error struct.
+      const errorMessageSizeBuffer = Buffer.from(
+        wasm.getMemorySlice(circuitFailureBufPtr + 2, circuitFailureBufPtr + 2 + 4),
+      );
+      const errorMessageSize = errorMessageSizeBuffer.readUint32BE();
+      // Now extract the entire `CircuitError` struct:
+      const errorBuf = Buffer.from(
+        wasm.getMemorySlice(circuitFailureBufPtr, circuitFailureBufPtr + 2 + 4 + errorMessageSize),
+      );
+      const err = CircuitError.fromBuffer(errorBuf);
+
+      // Free memory
+      wasm.call('bbfree', circuitFailureBufPtr);
+      wasm.call('bbfree', outputBufPtrPtr);
+      wasm.call('bbfree', inputBufPtr);
+
+      return Promise.reject(`Circuit failed with code ${err.code} - ${err.message}`);
```

Comment:
One minor tweak: let's throw an object that includes the code, so we don't need to regex it out of the message if we need to act on it from the caller. Maybe we can just `throw err` here, by having the `CircuitError` extend from `Error`...?

Reply:
On this: feel free to push it on another PR if you want to merge this one.